File: src/codes/forced-compatibility/ts/rfc_adapter.ts
// レガシーAPI（旧仕様）
class LegacyPrinter {
  printText(text: string): void {
    console.log(`*** ${text} ***`);
  }
}

// 新しいインターフェース
interface Printer {
  print(text: string): void;
}

// Adapter（旧APIをラップ）
class LegacyPrinterAdapter implements Printer {
  private legacy: LegacyPrinter;

  constructor() {
    this.legacy = new LegacyPrinter();
  }

  print(text: string): void {
    const formatted = text.toUpperCase(); // 必要な変換をここで吸収
    this.legacy.printText(formatted);
  }
}

// 新しいコード（Adapter に依存）
class ReportGenerator {
  constructor(private printer: Printer) {}

  printSummary(text: string): void {
    this.printer.print(text);
  }
}

// 利用例
const printer = new LegacyPrinterAdapter();
const report = new ReportGenerator(printer);
report.printSummary("月次レポート");

File: src/codes/forced-compatibility/py/rfc_adapter.py
from abc import ABC, abstractmethod

# レガシーAPI（旧仕様）
class LegacyPrinter:
    def print_text(self, text: str):
        print(f"*** {text} ***")

# 新しいインターフェース
class Printer(ABC):
    @abstractmethod
    def print(self, text: str):
        pass

# Adapter（旧APIをラップ）
class LegacyPrinterAdapter(Printer):
    def __init__(self):
        self.legacy = LegacyPrinter()

    def print(self, text: str):
        formatted = text.upper()
        self.legacy.print_text(formatted)

# 新しいコード
class ReportGenerator:
    def __init__(self, printer: Printer):
        self.printer = printer

    def print_summary(self, text: str):
        self.printer.print(text)

# 利用例
printer = LegacyPrinterAdapter()
report = ReportGenerator(printer)
report.print_summary("月次レポート")
